### PR TITLE
Cast to >=float32 tensor when passing scalar to self.log

### LINKED ISSUE #18984: cast to float32 or float64 tensor when passing scalar to self.log
### Bug description

https://github.com/Lightning-AI/lightning/pull/18686 made it such that all `ResultMetric` values are stored at `float32` or higher precision.

However, values passed in as floats to `self.log` still get auto-casted to a low precision type (if that's the default precision) via [`__to_tensor`](https://github.com/Lightning-AI/lightning/blob/1c86011dabaca211386be6acbee402c7bbda1f8c/src/lightning/pytorch/core/module.py#L454). This means that even though the internal result metric representation is precise, the values can loose precision before even reaching `ResultMetric`.

I'd suggest we modify `__to_tensor` to use a function like this (as introduced in https://github.com/Lightning-AI/lightning/pull/18686) to determine and set the dtype:

```py
def _get_default_dtype() -> torch.dtype:
    """The default dtype for new tensors, but no lower than float32."""
    dtype = torch.get_default_dtype()
    return dtype if dtype in (torch.float32, torch.float64) else torch.float32
```

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

_No response_

### Error messages and logs

N/A

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @carmocca

### PR DESCRIPTION
**Pull request by MF-FOOM**
## What does this PR do?

Ensures that values get cast to float32+ tensors when passed to `self.log`.

Fixes #18984

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [X] Is this pull request ready for review? (if not, please submit in draft mode)
- [X] Check that all items from **Before submitting** are resolved
- [X] Make sure the title is self-explanatory and the description concisely explains the PR
- [X] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19046.org.readthedocs.build/en/19046/

### CODE REVIEW

File: src/lightning/pytorch/core/module.py [hunk 1]
```diff
@@ -451,7 +451,7 @@ def log(
                 " but it should not contain information about `dataloader_idx`"
             )
 
-        value = apply_to_collection(value, (Tensor, numbers.Number), self.__to_tensor, name)
```
Comment (MF-FOOM) [inline src/lightning/pytorch/core/module.py:454]: Mote: I don't think `__to_tensor` is used anywhere else now? Should I remove it?

Comment (carmocca) [reply src/lightning/pytorch/core/module.py:454]: You could put your new implementation into `__to_tensor`, there's no need for a new method

Comment (MF-FOOM) [reply src/lightning/pytorch/core/module.py:454]: kk, was just worried it might surprise some folks

File: src/lightning/pytorch/core/module.py [hunk 2]
```diff
@@ -621,8 +621,18 @@ def __check_not_nested(value: dict, name: str) -> None:
     def __check_allowed(v: Any, name: str, value: Any) -> None:
         raise ValueError(f"`self.log({name}, {value})` was called, but `{type(v).__name__}` values cannot be logged")
 
+    @staticmethod
```
Comment (carmocca) [inline src/lightning/pytorch/core/module.py:624]: Can you instead import it from the logger connector?

```dif
diff --git a/tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py b/tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py
index bc2abdacb..2b88c283f 100644
--- a/tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py
+++ b/tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py
@@ -17,11 +17,13 @@ import collections
 import itertools
 from re import escape
 from unittest import mock
-from unittest.mock import call
+from unittest.mock import call, Mock
 
 import numpy as np
 import pytest
 import torch
+
+from lightning.fabric.plugins.precision.utils import _DtypeContextManager
 from lightning.pytorch import Trainer, callbacks
 from lightning.pytorch.callbacks import EarlyStopping, ModelCheckpoint, TQDMProgressBar
 from lightning.pytorch.core.module import LightningModule
@@ -779,3 +781,16 @@ def test_unsqueezed_tensor_logging():
     model.trainer = trainer
     model.log("foo", Tensor([1.2]), on_epoch=True)
     assert trainer.callback_metrics["foo"].ndim == 0
+
+
+def test_log_tensor_wrap_dtype():
+    class MyModel(LightningModule):
```

**Review by carmocca — APPROVED** [review]
I tried writing a test for this but it's complicated since the metric will be moved to `float32` after, so we would need to dynamically patch the call to `result.log`. It's not worth it.

```python
def test_log_tensor_wrap_dtype():
    class MyModel(LightningModule):
        def test_step(self, *_):
            self.log("foo", 1.23, batch_size=1)

    model = MyModel()
    trainer = Trainer(logger=False, enable_checkpointing=False, enable_model_summary=False, enable_progress_bar=False)
    # using the context manager class for cleanup, but users would use `set_default_dtype` directly
    with _DtypeContextManager(torch.float16):
        trainer.test(model, [0])
    assert trainer.callback_metrics["foo"].dtype == torch.float32
```

LGTM

File: src/lightning/pytorch/core/module.py (hunk 2 above)
Comment (MF-FOOM) [reply src/lightning/pytorch/core/module.py]: done!

**Review by awaelchli — APPROVED** [review]
